### PR TITLE
Globe fill extrusion

### PR DESCRIPTION
**Pull request by karimnaaji**
Add globe view fill-extrusion support, including:
  - Extended vertex array layout that gets only filled when globe projection is enabled
  - Polygon subdivision to match sphere surface as part of the fill-extrusion bucket processing step 
  - Render tests from gl-native

Screenshot from 2022-02-03 16-40-10

### CODE REVIEW

**Review by mpulkki-mapbox — COMMENTED** [review]
Looks great! Should we skip rendering of fill extrusion tiles if the data was created for a different projection? Following artefacts can be seen when zooming out quickly.

fill_extrusion_globe_zoom_out

**Comment by karimnaaji** [comment]
>Looks great! Should we skip rendering of fill extrusion tiles if the data was created for a different projection? Following artefacts can be seen when zooming out quickly.

Yes, Good catch! I'll fix that in that PR.

File: debug/globe-fill-extrusion.html [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+<script src='../dist/mapbox-gl-dev.js'></script>
+<script src='../debug/access_token_generated.js'></script>
+<script>
+
+var map = window.map = new mapboxgl.Map({
+    container: 'map',
+    zoom: 12.5,
+    center: [-122.4194, 37.7749],
+    style: 'mapbox://styles/endanke/ckw0r1ywi3f3b14mpoly55wjc',
```
[the remark this answers is not in the document]
Comment (avpeery) [reply debug/globe-fill-extrusion.html:25]: This seems fine for now and we can touch base again about getting a team account set up later!

**Review by mpulkki-mapbox — COMMENTED** [review]
I copied the style from `debug/globe-fill-extrusion.html` to `debug/terrain-debug.html` and noticed that there's no longer elevation differences between the countries and only single dark gray color is used. Also in lower zoom levels some countries are either missing or flickering when the terrain is toggled on.

I believe that at least missing countries and fill extrusion flickering is caused by some issues in the flat roof computations when the terrain is enabled. Similar issues have been fixed few times before in gl-native but I'm not sure if those were ported to gl-js. cc @astojilj 

globe_fill_extrusion_countries_missing

**Comment by karimnaaji** [comment]
Looking at that. I wonder if it's related to the define `HAS_CENTROID` that isn't enabled in gl-js (from https://github.com/mapbox/mapbox-gl-js/pull/10986/).

**Comment by karimnaaji** [comment]
@mpulkki-mapbox The color issue is incoming from the example itself, as it overrides any `fill-extrusion` properties, including colors, [here](https://github.com/mapbox/mapbox-gl-js/blob/main/debug/terrain-debug.html#L111_L128). I added an option for terrain in the new example, where this issue isn't visible: https://github.com/mapbox/mapbox-gl-js/pull/11445/commits/46c92af25b0733fddabdd8eb489dd8c852e7bf45.

Concerning the missing polygons, it's because the flat-roof code path was enabled with globe when terrain was set and the centroid buffer bound. I pushed a fix in https://github.com/mapbox/mapbox-gl-js/pull/11445/commits/dac54a1c9cff1666cb9ebf62cc7d193d90765414 to disable centroid buffer as flat roofs should be mutually exclusive with globe-view.

I also noticed it's an issue in native, note the missing polygon for Russia, let me know if this should be back-ported there so we can track that fix:

Screen Shot 2022-02-15 at 12 47 05 PM

**Comment by mpulkki-mapbox** [comment]
> @mpulkki-mapbox The color issue is incoming from the example itself, as it overrides any `fill-extrusion` properties, including colors, [here](https://github.com/mapbox/mapbox-gl-js/blob/main/debug/terrain-debug.html#L111_L128). I added an option for terrain in the new example, where this issue isn't visible: [46c92af](https://github.com/mapbox/mapbox-gl-js/commit/46c92af25b0733fddabdd8eb489dd8c852e7bf45).

Thanks for checking the issue! I actually dug a bit deeper myself too and found out the style to have two separate fill extrusion layers named `country-boundaries` and `country-boundaries (1)`. The latter one is the dark one and it is set to visible in `terrain-debug.html`
```javascript
document.getElementById('satellite-checkbox').onclick = function() {
    for (const layerId in map.style._layers) {
        if (layerId.startsWith('satellite') || layerId === buildingsLayerId) continue;
        map.setLayoutProperty(layerId, 'visibility', !this.checked ? 'visible' : 'none');
    }
    map.setLayoutProperty('satellite', 'visibility', this.checked ? 'visible' : 'none');
    map.setLayoutProperty('satellite-background', 'visibility', this.checked ? 'visible' : 'none');
    document.getElementById('road-label-checkbox').onclick();
};
```

I believe there's some issues left with the flat roofs. The following assert can be triggered in both `globe-fill-extrusion.html` and `debug-terrain.html` by first enabling the terrain and then disabling the globe (I just set the projection to null in chrome's console).

image

**Comment by karimnaaji** [comment]
>I believe there's some issues left with the flat roofs. The following assert can be >triggered in both globe-fill-extrusion.html and debug-terrain.html by first enabling the >terrain and then disabling the globe (I just set the projection to null in chrome's >console).
>
>EDIT: Just launch terrain-debug.html with globe disabled by default will lead to this assertation.

@mpulkki-mapbox thanks for testing that thoroughly, after more testing on my side I've been narrowing it down to this more minimal case (a location that triggers it is `#2.64/50.56/32.43/0/4`):
```html
<!DOCTYPE html>
<html>
<head>
    <title>Mapbox GL JS debug page</title>
    <meta charset='utf-8'>
    <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=no">
    <link rel='stylesheet' href='../dist/mapbox-gl.css' />
    <style>
        body { margin: 0; padding: 0; }
        html, body, #map { height: 100%; }

        #checkboxes {
            position: absolute;
            top:0;
            left:0;
            padding:10px;
        }
    </style>
</head>

<body>
<div id='map'></div>

<script src='../dist/mapbox-gl-dev.js'></script>
<script src='../debug/access_token_generated.js'></script>
<script>

var map = window.map = new mapboxgl.Map({
    container: 'map',
    zoom: 12.5,
    center: [-122.4194, 37.7749],
    style: 'mapbox://styles/endanke/ckw0r1ywi3f3b14mpoly55wjc',
    hash: true
});

map.on('style.load', function() {
    map.addSource('mapbox-dem', {
        "type": "raster-dem",
        "url": "mapbox://mapbox.terrain-rgb",
        "tileSize": 512,
        "maxzoom": 14
    });
    map.setTerrain({"source": "mapbox-dem"});
});

</script>
</body>
</html>
```
Which is basically the combination of this complex `fill-extrusion` style along with 3d terrain without globe. Disabling or enabling globe isn't directly related to this issue, as starting out with Mercator also triggers these assertions. I've opened a ticket to follow-up on that issue and track it in https://github.com/mapbox/mapbox-gl-js/issues/11516.

**Review by mpulkki-mapbox — APPROVED** [review]

**Pull request merged**